File: two_sum.java
/*

1. Two Sum

Question: https://leetcode.com/problems/two-sum/

*/


/* 

Easy to understand

*/

public class Solution {
    public int[] twoSum(int[] nums, int target) {
        
        int[] indices = new int[2];
        
        for( int i = 0; i < nums.length; i++ ){
            for(int j = i+1; j < nums.length; j++){
                if(nums[i]+nums[j] == target){
                    indices[0] = i;
                    indices[1] = j;
                    break;
                }
            }
        }
        
        return indices;
    }
}


/*

Optimised solution

*/


public class Solution {
    public int[] twoSum(int[] nums, int target) {
        
        HashMap<Integer,Integer> hm = new HashMap<Integer,Integer>();
        
        int[] indices = new int[2];
        
        for( int i = 0; i < nums.length; i++ ){
            Integer val = hm.get(target-nums[i]);
            if( val != null ){
                indices[0] = val;
                indices[1] = i;
                break;
            }
            hm.put(nums[i],i);
        }
        
        return indices;
    }
}

File: two_sum.py
"""

1. Two Sum

Question: https://leetcode.com/problems/two-sum/

"""

"""
Easy to understand solution
"""

class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        for i in xrange(len(nums)):
            for j in xrange(i+1,len(nums)):
                if nums[i]+nums[j] == target:
                    return [i,j]

"""
Optimised solution
"""
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        dic = {}
        for i in range(len(nums)):
            if (target - nums[i]) in dic:
                return [dic[target-nums[i]],i]
            dic[nums[i]] = i